### PR TITLE
Adding AES_GCM encryption mode to overcome AES_CBC limitations

### PR DESCRIPTION
**Pull request by anshubit5516**
* **Problem Overview**  
  Currently Concord supports AES_CBC as mode of encryption, which is vulnerable to oracle padding attack. More information can be found here(https://en.wikipedia.org/wiki/Padding_oracle_attack).
So added AES_GCM mode for all encryption/decryption functionalities. GCM is more secured and stable.
* **Testing Done**  
  Unit testing is done and added in the PR.

### CODE REVIEW

File: libs/secrets/src/aes.cpp
Comment (eyalrund) [inline libs/secrets/src/aes.cpp:45]: why is it using aes_256_cbc?

File: libs/secrets/src/secrets_manager_enc.cpp
Comment (eyalrund) [inline libs/secrets/src/secrets_manager_enc.cpp:33]: maybe call it getAESEncryptionMode()

File: libs/secrets/src/aes.cpp
Comment (eyalrund) [inline libs/secrets/src/aes.cpp:27]: Why is edDSA part of this class?

**Review by eyalrund — CHANGES_REQUESTED** [review]

**Comment by eyalrund** [comment]
Testing must be done also with real deployment

File: libs/secrets/src/aes.h
Comment (cjn) [inline libs/secrets/src/aes.h:47]: Do we need algo string ? If so is this the correct default value for the interface?

File: libs/secrets/src/aes.h
Comment (cjn) [inline libs/secrets/src/aes.h:44]: iAES_Mode - Can you check if we use this naming convention for other abstract classes?

**Review by toly-kournik — CHANGES_REQUESTED** [review]
@anshubit5516,
this change should enter our `concord-crypto` library which will shortly spin off the `util` library.
In any case there shouldn't be any direct usage of `OpenSSL` library except from there.
This allows for seamless switching of underlying crypto implementations.
My suggestion is to wait a bit until we merge the `concord-crypto` separation merge this week

File: libs/secrets/src/aes.cpp
Comment (anshubit5516) [reply libs/secrets/src/aes.cpp:45]: Since we are supporting both (CBC and GCM) and this method is CBC implementation.

File: libs/secrets/src/secrets_manager_enc.cpp
Comment (anshubit5516) [reply libs/secrets/src/secrets_manager_enc.cpp:33]: done

File: libs/secrets/src/aes.h
Comment (anshubit5516) [reply libs/secrets/src/aes.h:44]: What i have seen in the code, generally interface names are like IVerifier,ISigner..
I can rename the interface as IAesMode just to be aligned.

File: libs/secrets/src/aes.cpp
Comment (anshubit5516) [reply libs/secrets/src/aes.cpp]: removed

File: libs/secrets/src/aes.h
Comment (anshubit5516) [reply libs/secrets/src/aes.h]: i confirmed with code owners and removed it.

**Comment by anshubit5516** [comment]
> @anshubit5516, this change should enter our `concord-crypto` library which will shortly spin off the `util` library. In any case there shouldn't be any direct usage of `OpenSSL` library except from there. This allows for seamless switching of underlying crypto implementations. My suggestion is to wait a bit until we merge the `concord-crypto` separation merge this week

As discussed, i will get this PR merged and then after crypto folder is organised, will move the code

**Review by eyalrund — APPROVED** [review]

**Comment by cjn** [comment]
LGTM. Thanks @anshubit5516

**Review by cjn — APPROVED** [review]

**Review by toly-kournik — APPROVED** [review]
LGTM

**Pull request merged**
